build.assets/robotest/Makefile

```diff
@@ -0,0 +1,326 @@
+
+.PHONY: images-upgrade
+images-upgrade: ## Create/fetch robotest cluster images needed for upgrade testing.
+images-upgrade:
+	$(MAKE) -j $(ROBOTEST_UPGRADE_IMAGES)
+
+.PHONY: image-opscenter
+image-opscenter: ## Move opscenter.tar cluster image into place for test execution.
+image-opscenter: $(ROBOTEST_OPSCENTER_IMAGE)
+
+$(ROBOTEST_OPSCENTER_IMAGE): $(OPSCENTER_OUT) | $(ROBOTEST_IMAGEDIR)
+	cp $(OPSCENTER_OUT) $(ROBOTEST_OPSCENTER_IMAGE)
+
+# The following rule populates the build specific docker image mount
+# from cached images. Fulfills the $(ROBOTEST_UPGRADE_IMAGES) targets.
+# .PRECIOUS because these need to be present through the run target.
+.PRECIOUS: $(ROBOTEST_IMAGEDIR)/robotest-%.tar
+$(ROBOTEST_IMAGEDIR)/robotest-%.tar: $(ROBOTEST_UPGRADE_IMAGE_CACHE)/robotest-%.tar | $(ROBOTEST_IMAGEDIR) $(ROBOTEST_BUILD_TMP)
+	$(eval TMP := $(shell mktemp -d -p $(ROBOTEST_BUILD_TMP)))
+	$(eval ATOMIC := $(TMP)/robotest-$*.tar)
+	cp $< $(ATOMIC)
+	mv $(ATOMIC) $@
+	rm -rf $(TMP)
+
+# The following target is a pattern rule for populating the upgrade image cache
+# Fulfills the $(ROBOTEST_UPGRADE_CACHE_IMAGES) targets.
+# .PRECIOUS because it is a cache shared across versions.
+.PRECIOUS: $(ROBOTEST_UPGRADE_IMAGE_CACHE)/robotest-%.tar
+$(ROBOTEST_UPGRADE_IMAGE_CACHE)/robotest-%.tar: $(ROBOTEST_TELE_CACHE)/tele-% $(ROBOTEST_UPGRADE_BASE_IMAGE_SRC) | $(ROBOTEST_UPGRADE_IMAGE_CACHE) $(ROBOTEST_CACHE_TMP)
+	$(eval TMP := $(shell mktemp -d -p $(ROBOTEST_CACHE_TMP)))
+	$(eval TGT := $(TMP)/robotest-$*.tar)
+	$< build \
+		$(ROBOTEST_UPGRADE_BASE_APP_YAML) \
+		--state-dir $(ROBOTEST_GRAVITY_PKG_CACHE) \
+		--version $* \
+		--output=$(TGT)
+	mv $(TGT) $@
+	rm -rf $(TMP)
+
+
+# The following target is a pattern rule for populating the tele binary cache.
+# Fulfills the $(ROBOTEST_TELE_BINARIES) targets.
+# .PRECIOUS because it is a cache shared across branches/builds.
+.PRECIOUS: $(ROBOTEST_TELE_CACHE)/tele-%
+$(ROBOTEST_TELE_CACHE)/tele-%: | $(ROBOTEST_CACHE_TMP) $(ROBOTEST_TELE_CACHE)
+	$(eval TMP := $(shell mktemp -d -p $(ROBOTEST_CACHE_TMP)))
+ifdef ENTERPRISE
+	wget --quiet https://get.gravitational.io/telekube/bin/$*/linux/x86_64/tele --output-document $(TMP)/tele
+	chmod u+x $(TMP)/tele
+else
+	wget --quiet https://get.gravitational.com/gravity-$*-linux-x86_64-bin.tar.gz --directory-prefix $(TMP)
+	tar -xzf $(TMP)/gravity-$*-linux-x86_64-bin.tar.gz -C $(TMP)
+endif
+	mv $(TMP)/tele $@
+	rm -rf $(TMP)
+
+.PHONY: download-teles
+download-teles: ## Download tele binaries needed to build robotest images.
+download-teles:
+	$(MAKE) -j $(ROBOTEST_TELE_BINARIES)
+
+.PHONY: populate-caches
+populate-caches: ## Populate caches with tele binaries, gravity packages, and robotest upgrade images.
+	$(MAKE) -j $(ROBOTEST_TELE_BINARIES)
+	$(MAKE) -j $(ROBOTEST_UPGRADE_CACHE_IMAGES)
+
+.PHONY: get-upgrade-from-versions
+get-upgrade-from-versions: ## Show versions used in upgrade from testing.
+get-upgrade-from-versions:
+	@echo $(ROBOTEST_UPGRADE_VERSIONS)
```

Comment:
This should probably be `get config` and print out the rendered config.  I'm not sure there is much reason to care about the upgrade versions alone (aside from those hacking on this file).